### PR TITLE
UCP/UCT: Pack both iface and ep addrs in wireup

### PR DESCRIPTION
**Pull request by brminich**
Iface address is always needed to correctly check peer reachability (even if ep address is present).
The check is done by `uct_iface_is_reachable` in `ucp_wireup_is_reachable`

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4112/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4116/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@yosefe, pls review

File: src/ucp/wireup/address.c [hunk 1]
```diff
@@ -563,16 +575,19 @@ ucs_status_t ucp_address_unpack(const void *buffer, uint64_t *remote_uuid_p,
...
-            address->tl_addr      = (tl_addr_len > 0) ? ptr : NULL;
-            address->tl_addr_len  = tl_addr_len;
+            address->dev_addr   = (dev_addr_len > 0) ? dev_addr : NULL;
+            address->md_index   = md_index;
+            address->md_flags   = md_flags;
+            address->iface_addr = (iface_addr_len > 0) ? ptr : NULL;
+
+            ptr                += iface_addr_len;
+            ep_addr_len         = *(uint8_t*)ptr;
+            address->ep_addr    = ep_addr_len ? ptr + 1 : NULL;
```
Comment (yosefe) [inline src/ucp/wireup/address.c]: (ep_addr_len > 0) for consistency with line 585?

File: src/ucp/wireup/address.h [hunk 2]
```diff
@@ -53,17 +53,12 @@ struct ucp_address_iface_attr {
...
     ucp_rsc_index_t            md_index;       /* Memory domain index */
     uint64_t                   md_flags;       /* MD reg/alloc flags */
     ucp_address_iface_attr_t   iface_attr;     /* Interface attributes information */
-    size_t                     tl_addr_len;    /* Transport address length */
-    union {
-        const uct_iface_addr_t *iface_addr;    /* Interface address */
-        const uct_ep_addr_t    *ep_addr;       /* Endpoint address */
-        const void             *tl_addr;
-    };
+    const uct_iface_addr_t     *iface_addr;    /* Interface address */
```
Comment (yosefe) [inline src/ucp/wireup/address.h]: /* Interface address, NULL if not available */

File: src/ucp/wireup/address.h [hunk 3]
```diff
@@ -53,17 +53,12 @@ struct ucp_address_iface_attr {
...
+    const uct_ep_addr_t        *ep_addr;       /* Endpoint address */
```
Comment (yosefe) [inline src/ucp/wireup/address.h]: ... NULL if not available */

File: src/ucp/wireup/address.c [hunk 4]
```diff
@@ -137,21 +137,21 @@ ucp_address_gather_devices(ucp_worker_h worker, uint64_t tl_bitmap, int has_ep,
...
+        dev->tl_addrs_size += iface_attr->iface_addr_len;
+
+        if (!(iface_attr->cap.flags & UCT_IFACE_FLAG_CONNECT_TO_IFACE) && has_ep) {
+            dev->tl_addrs_size += iface_attr->ep_addr_len;
+        }
+
         dev->tl_addrs_size += sizeof(uint16_t); /* tl name checksum */
         dev->tl_addrs_size += sizeof(ucp_address_packed_iface_attr_t); /* iface attr */
-        dev->tl_addrs_size += 1;                /* address length */
+        dev->tl_addrs_size += 2;                /* iface and ep address lengths */
```
Comment (yosefe) [inline src/ucp/wireup/address.c]: can add flag for tl_addr_len, in addition to UCP_ADDRESS_FLAG_LAST, which would indicate if ep address is present. this way no need to increase address size for modex

**Comment by yosefe** [comment]
also - need to remove " " from commit message

**Comment by MattBBaker** [comment]
Change tested on Titan, wireup works.

**Comment by brminich** [comment]
thanks @MattBBaker!

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4122/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4123/ for details (Mellanox internal link).

**Pull request merged**